File: packages/r/src/rust/src/options.rs
//! Option decoding for R bindings.

use extendr_api::prelude::*;

/// Helper: extract and convert a value from an R list by name.
fn list_get(list: &List, key: &str) -> Option<Robj> {
    let names = list.names().ok();
    names
        .iter()
        .zip(list.iter())
        .find(|(name, _)| name == key)
        .map(|(_, val)| val)
}

/// Decode a preprocessing preset enum from its string representation.
fn decode_preprocessing_preset(val: Robj) -> std::result::Result<crate::PreprocessingPreset, String> {
    let s = String::try_from(&val).map_err(|e| format!("preprocessing_preset: {e}"))?;
    match s.as_str() {
        "Minimal" => Ok(crate::PreprocessingPreset::Minimal),
        "Standard" => Ok(crate::PreprocessingPreset::Standard),
        "Aggressive" => Ok(crate::PreprocessingPreset::Aggressive),
        _ => Err(format!("preprocessing_preset: unknown variant '{{}}'", s)),
    }
}

/// Decode a list indent type enum from its string representation.
fn decode_list_indent_type(val: Robj) -> std::result::Result<crate::ListIndentType, String> {
    let s = String::try_from(&val).map_err(|e| format!("list_indent_type: {e}"))?;
    match s.as_str() {
        "Spaces" => Ok(crate::ListIndentType::Spaces),
        "Tabs" => Ok(crate::ListIndentType::Tabs),
        _ => Err(format!("list_indent_type: unknown variant '{{}}'", s)),
    }
}

/// Decode a whitespace mode enum from its string representation.
fn decode_whitespace_mode(val: Robj) -> std::result::Result<crate::WhitespaceMode, String> {
    let s = String::try_from(&val).map_err(|e| format!("whitespace_mode: {e}"))?;
    match s.as_str() {
        "Normalized" => Ok(crate::WhitespaceMode::Normalized),
        "Strict" => Ok(crate::WhitespaceMode::Strict),
        _ => Err(format!("whitespace_mode: unknown variant '{{}}'", s)),
    }
}

/// Decode a url escape style enum from its string representation.
fn decode_url_escape_style(val: Robj) -> std::result::Result<crate::UrlEscapeStyle, String> {
    let s = String::try_from(&val).map_err(|e| format!("url_escape_style: {e}"))?;
    match s.as_str() {
        "Angle" => Ok(crate::UrlEscapeStyle::Angle),
        "Percent" => Ok(crate::UrlEscapeStyle::Percent),
        _ => Err(format!("url_escape_style: unknown variant '{{}}'", s)),
    }
}

/// Decode a highlight style enum from its string representation.
fn decode_highlight_style(val: Robj) -> std::result::Result<crate::HighlightStyle, String> {
    let s = String::try_from(&val).map_err(|e| format!("highlight_style: {e}"))?;
    match s.as_str() {
        "DoubleEqual" => Ok(crate::HighlightStyle::DoubleEqual),
        "Html" => Ok(crate::HighlightStyle::Html),
        "Bold" => Ok(crate::HighlightStyle::Bold),
        "None" => Ok(crate::HighlightStyle::None),
        _ => Err(format!("highlight_style: unknown variant '{{}}'", s)),
    }
}

/// Decode a link style enum from its string representation.
fn decode_link_style(val: Robj) -> std::result::Result<crate::LinkStyle, String> {
    let s = String::try_from(&val).map_err(|e| format!("link_style: {e}"))?;
    match s.as_str() {
        "Inline" => Ok(crate::LinkStyle::Inline),
        "Reference" => Ok(crate::LinkStyle::Reference),
        _ => Err(format!("link_style: unknown variant '{{}}'", s)),
    }
}

/// Decode a newline style enum from its string representation.
fn decode_newline_style(val: Robj) -> std::result::Result<crate::NewlineStyle, String> {
    let s = String::try_from(&val).map_err(|e| format!("newline_style: {e}"))?;
    match s.as_str() {
        "Spaces" => Ok(crate::NewlineStyle::Spaces),
        "Backslash" => Ok(crate::NewlineStyle::Backslash),
        _ => Err(format!("newline_style: unknown variant '{{}}'", s)),
    }
}

/// Decode a heading style enum from its string representation.
fn decode_heading_style(val: Robj) -> std::result::Result<crate::HeadingStyle, String> {
    let s = String::try_from(&val).map_err(|e| format!("heading_style: {e}"))?;
    match s.as_str() {
        "Atx" => Ok(crate::HeadingStyle::Atx),
        "Underlined" => Ok(crate::HeadingStyle::Underlined),
        "AtxClosed" => Ok(crate::HeadingStyle::AtxClosed),
        _ => Err(format!("heading_style: unknown variant '{{}}'", s)),
    }
}

/// Decode a code block style enum from its string representation.
fn decode_code_block_style(val: Robj) -> std::result::Result<crate::CodeBlockStyle, String> {
    let s = String::try_from(&val).map_err(|e| format!("code_block_style: {e}"))?;
    match s.as_str() {
        "Indented" => Ok(crate::CodeBlockStyle::Indented),
        "Backticks" => Ok(crate::CodeBlockStyle::Backticks),
        "Tildes" => Ok(crate::CodeBlockStyle::Tildes),
        _ => Err(format!("code_block_style: unknown variant '{{}}'", s)),
    }
}

/// Decode a output format enum from its string representation.
fn decode_output_format(val: Robj) -> std::result::Result<crate::OutputFormat, String> {
    let s = String::try_from(&val).map_err(|e| format!("output_format: {e}"))?;
    match s.as_str() {
        "Markdown" => Ok(crate::OutputFormat::Markdown),
        "Djot" => Ok(crate::OutputFormat::Djot),
        "Plain" => Ok(crate::OutputFormat::Plain),
        _ => Err(format!("output_format: unknown variant '{{}}'", s)),
    }
}

/// Decode preprocessing options from an R list.
fn decode_preprocessing_options(val: Robj) -> std::result::Result<crate::PreprocessingOptions, String> {
    if val.is_null() {
        return Ok(crate::PreprocessingOptions::default());
    }
    let list = List::try_from(&val).map_err(|e| format!("preprocessing: {e}"))?;
    let mut opts = crate::PreprocessingOptions::default();

    if let Some(v) = list_get(&list, "enabled") {
        opts.enabled = bool::try_from(&v).map_err(|e| format!("preprocessing.enabled: {e}"))?;
    }
    if let Some(v) = list_get(&list, "preset") {
        opts.preset = decode_preprocessing_preset(v)?;
    }
    if let Some(v) = list_get(&list, "remove_navigation") {
        opts.remove_navigation = bool::try_from(&v).map_err(|e| format!("preprocessing.remove_navigation: {e}"))?;
    }
    if let Some(v) = list_get(&list, "remove_forms") {
        opts.remove_forms = bool::try_from(&v).map_err(|e| format!("preprocessing.remove_forms: {e}"))?;
    }

    Ok(opts)
}

/// Decode an R ExternalPtr, NULL, or named list into ConversionOptions.
///
/// Accepts:
/// - ExternalPtr<ConversionOptions> (from $default() or builder methods) — unwraps and converts
/// - NULL — returns default ConversionOptions
/// - Named list with field names matching struct fields — decodes field by field
///
/// Fields are optional: omitted fields retain their defaults. Unknown fields are ignored.
pub fn decode_options(options: Robj) -> std::result::Result<crate::ConversionOptions, String> {
    if options.is_null() {
        return Ok(crate::ConversionOptions::default());
    }

    // Accept the wrapper struct returned by `ConversionOptions$default()` / builder methods,
    // which extendr exposes as an `ExternalPtr`. The binding struct is returned directly
    // from the #[extendr] impl methods, so unwrap it as the binding type.
    if let Ok(ext) = ExternalPtr::<crate::ConversionOptions>::try_from(&options) {
        // Clone the binding struct and convert to core type via the generated From impl
        return Ok((*ext).clone().into());
    }

    // Try to decode as a named list
    let list =
        List::try_from(&options).map_err(|e| format!("options must be NULL, ExternalPtr, or named list: {e}"))?;
    let mut opts = crate::ConversionOptions::default();

    if let Some(v) = list_get(&list, "heading_style") {
        opts.heading_style = decode_heading_style(v)?;
    }
    if let Some(v) = list_get(&list, "list_indent_type") {
        opts.list_indent_type = decode_list_indent_type(v)?;
    }
    if let Some(v) = list_get(&list, "list_indent_width") {
        opts.list_indent_width = f64::try_from(&v).map_err(|e| format!("list_indent_width: {e}"))?;
    }
    if let Some(v) = list_get(&list, "bullets") {
        opts.bullets = String::try_from(&v).map_err(|e| format!("bullets: {e}"))?;
    }
    if let Some(v) = list_get(&list, "strong_em_symbol") {
        opts.strong_em_symbol = String::try_from(&v).map_err(|e| format!("strong_em_symbol: {e}"))?;
    }
    if let Some(v) = list_get(&list, "escape_asterisks") {
        opts.escape_asterisks = bool::try_from(&v).map_err(|e| format!("escape_asterisks: {e}"))?;
    }
    if let Some(v) = list_get(&list, "escape_underscores") {
        opts.escape_underscores = bool::try_from(&v).map_err(|e| format!("escape_underscores: {e}"))?;
    }
    if let Some(v) = list_get(&list, "escape_misc") {
        opts.escape_misc = bool::try_from(&v).map_err(|e| format!("escape_misc: {e}"))?;
    }
    if let Some(v) = list_get(&list, "escape_ascii") {
        opts.escape_ascii = bool::try_from(&v).map_err(|e| format!("escape_ascii: {e}"))?;
    }
    if let Some(v) = list_get(&list, "code_language") {
        opts.code_language = String::try_from(&v).map_err(|e| format!("code_language: {e}"))?;
    }
    if let Some(v) = list_get(&list, "autolinks") {
        opts.autolinks = bool::try_from(&v).map_err(|e| format!("autolinks: {e}"))?;
    }
    if let Some(v) = list_get(&list, "default_title") {
        opts.default_title = bool::try_from(&v).map_err(|e| format!("default_title: {e}"))?;
    }
    if let Some(v) = list_get(&list, "br_in_tables") {
        opts.br_in_tables = bool::try_from(&v).map_err(|e| format!("br_in_tables: {e}"))?;
    }
    if let Some(v) = list_get(&list, "compact_tables") {
        opts.compact_tables = bool::try_from(&v).map_err(|e| format!("compact_tables: {e}"))?;
    }
    if let Some(v) = list_get(&list, "highlight_style") {
        opts.highlight_style = decode_highlight_style(v)?;
    }
    if let Some(v) = list_get(&list, "extract_metadata") {
        opts.extract_metadata = bool::try_from(&v).map_err(|e| format!("extract_metadata: {e}"))?;
    }
    if let Some(v) = list_get(&list, "whitespace_mode") {
        opts.whitespace_mode = decode_whitespace_mode(v)?;
    }
    if let Some(v) = list_get(&list, "strip_newlines") {
        opts.strip_newlines = bool::try_from(&v).map_err(|e| format!("strip_newlines: {e}"))?;
    }
    if let Some(v) = list_get(&list, "wrap") {
        opts.wrap = bool::try_from(&v).map_err(|e| format!("wrap: {e}"))?;
    }
    if let Some(v) = list_get(&list, "wrap_width") {
        opts.wrap_width = f64::try_from(&v).map_err(|e| format!("wrap_width: {e}"))?;
    }
    if let Some(v) = list_get(&list, "convert_as_inline") {
        opts.convert_as_inline = bool::try_from(&v).map_err(|e| format!("convert_as_inline: {e}"))?;
    }
    if let Some(v) = list_get(&list, "sub_symbol") {
        opts.sub_symbol = String::try_from(&v).map_err(|e| format!("sub_symbol: {e}"))?;
    }
    if let Some(v) = list_get(&list, "sup_symbol") {
        opts.sup_symbol = String::try_from(&v).map_err(|e| format!("sup_symbol: {e}"))?;
    }
    if let Some(v) = list_get(&list, "newline_style") {
        opts.newline_style = decode_newline_style(v)?;
    }
    if let Some(v) = list_get(&list, "code_block_style") {
        opts.code_block_style = decode_code_block_style(v)?;
    }
    if let Some(v) = list_get(&list, "keep_inline_images_in") {
        let strings = Strings::try_from(&v).map_err(|e| format!("keep_inline_images_in: {e}"))?;
        let vec: Vec<String> = strings.iter().map(|s| s.to_string()).collect();
        opts.keep_inline_images_in = vec;
    }
    if let Some(v) = list_get(&list, "preprocessing") {
        opts.preprocessing = decode_preprocessing_options(v)?;
    }
    if let Some(v) = list_get(&list, "encoding") {
        opts.encoding = String::try_from(&v).map_err(|e| format!("encoding: {e}"))?;
    }
    if let Some(v) = list_get(&list, "debug") {
        opts.debug = bool::try_from(&v).map_err(|e| format!("debug: {e}"))?;
    }
    if let Some(v) = list_get(&list, "strip_tags") {
        let strings = Strings::try_from(&v).map_err(|e| format!("strip_tags: {e}"))?;
        let vec: Vec<String> = strings.iter().map(|s| s.to_string()).collect();
        opts.strip_tags = vec;
    }
    if let Some(v) = list_get(&list, "preserve_tags") {
        let strings = Strings::try_from(&v).map_err(|e| format!("preserve_tags: {e}"))?;
        let vec: Vec<String> = strings.iter().map(|s| s.to_string()).collect();
        opts.preserve_tags = vec;
    }
    if let Some(v) = list_get(&list, "skip_images") {
        opts.skip_images = bool::try_from(&v).map_err(|e| format!("skip_images: {e}"))?;
    }
    if let Some(v) = list_get(&list, "url_escape_style") {
        opts.url_escape_style = decode_url_escape_style(v)?;
    }
    if let Some(v) = list_get(&list, "link_style") {
        opts.link_style = decode_link_style(v)?;
    }
    if let Some(v) = list_get(&list, "output_format") {
        opts.output_format = decode_output_format(v)?;
    }
    if let Some(v) = list_get(&list, "include_document_structure") {
        opts.include_document_structure = bool::try_from(&v).map_err(|e| format!("include_document_structure: {e}"))?;
    }
    if let Some(v) = list_get(&list, "extract_images") {
        opts.extract_images = bool::try_from(&v).map_err(|e| format!("extract_images: {e}"))?;
    }
    if let Some(v) = list_get(&list, "max_image_size") {
        opts.max_image_size = f64::try_from(&v).map_err(|e| format!("max_image_size: {e}"))?;
    }
    if let Some(v) = list_get(&list, "capture_svg") {
        opts.capture_svg = bool::try_from(&v).map_err(|e| format!("capture_svg: {e}"))?;
    }
    if let Some(v) = list_get(&list, "infer_dimensions") {
        opts.infer_dimensions = bool::try_from(&v).map_err(|e| format!("infer_dimensions: {e}"))?;
    }
    if let Some(v) = list_get(&list, "max_depth") {
        opts.max_depth = f64::try_from(&v).map_err(|e| format!("max_depth: {e}"))?;
    }
    if let Some(v) = list_get(&list, "exclude_selectors") {
        let strings = Strings::try_from(&v).map_err(|e| format!("exclude_selectors: {e}"))?;
        let vec: Vec<String> = strings.iter().map(|s| s.to_string()).collect();
        opts.exclude_selectors = vec;
    }
    // Note: visitor field is skipped — R has no visitor concept, so it remains at default None

    Ok(opts)
}

File: packages/r/R/options.R
#' Create a ConversionOptions list for generated bindings
#'
#' All parameters default to `NULL`, which means the Rust default is used.
#' Pass named arguments to override individual settings.
#'
#' @param heading_style Heading style to use in Markdown output (ATX `#` or Setext underline)
#' @param list_indent_type How to indent nested list items (spaces or tab)
#' @param list_indent_width Number of spaces (or tabs) to use for each level of list indentation
#' @param bullets Bullet character(s) to use for unordered list items (e.g. `"-"`, `"*"`)
#' @param strong_em_symbol Character used for bold/italic emphasis markers (`*` or `_`)
#' @param escape_asterisks Escape `*` characters in plain text to avoid unintended bold/italic
#' @param escape_underscores Escape `_` characters in plain text to avoid unintended bold/italic
#' @param escape_misc Escape miscellaneous Markdown metacharacters (`[]()#` etc.) in plain text
#' @param escape_ascii Escape ASCII characters that have special meaning in certain Markdown dialects
#' @param code_language Default language annotation for fenced code blocks that have no language hint
#' @param autolinks Automatically convert bare URLs into Markdown autolinks
#' @param default_title Emit a default title when no `<title>` tag is present
#' @param br_in_tables Render `<br>` elements inside table cells as literal line breaks
#' @param compact_tables Emit tables without column padding (compact GFM format)
#' @param highlight_style Style used for `<mark>` / highlighted text (e.g. `==text==`)
#' @param extract_metadata Populate `result.metadata` with `<head>` / `<meta>` extraction
#' @param whitespace_mode Controls how whitespace sequences are normalised in the converted output
#' @param strip_newlines Strip all newlines from the output, producing a single-line result
#' @param wrap Wrap long lines at [`wrap_width`](Self::wrap_width) characters
#' @param wrap_width Maximum output line width in characters when [`wrap`](Self::wrap) is `true` (default `80`)
#' @param convert_as_inline Treat the entire document as inline content (no block-level wrappers)
#' @param sub_symbol Markdown notation for subscript text (e.g. `"~"`)
#' @param sup_symbol Markdown notation for superscript text (e.g. `"^"`)
#' @param newline_style How to encode hard line breaks (`<br>`) in Markdown
#' @param code_block_style Style used for fenced code blocks (backticks or tilde)
#' @param keep_inline_images_in HTML tag names whose `<img>` children are kept inline instead of block
#' @param preprocessing Options for the HTML pre-processing pass applied before conversion begins
#' @param encoding Expected character encoding of the input HTML (default `"utf-8"`)
#' @param debug Emit debug information during conversion
#' @param strip_tags HTML tag names whose content is stripped from the output entirely
#' @param preserve_tags HTML tag names that are preserved verbatim in the output
#' @param skip_images Skip conversion of `<img>` elements (omit images from output)
#' @param url_escape_style URL encoding strategy for link and image destinations
#' @param link_style Link rendering style (inline or reference)
#' @param output_format Target output format (Markdown, plain text, etc.)
#' @param include_document_structure Include structured document tree in result
#' @param extract_images Extract inline images from data URIs and SVGs
#' @param max_image_size Maximum decoded image size in bytes (default 5MB)
#' @param capture_svg Capture SVG elements as images
#' @param infer_dimensions Infer image dimensions from data
#' @param max_depth Maximum DOM traversal depth. `None` means unlimited
#' @param exclude_selectors CSS selectors for elements to exclude entirely (element + all content)
#' @param tier_strategy Which conversion tier to use
#' @param visitor (feature-gated) Optional visitor for custom traversal logic
#' @return A named list suitable for the `options` argument of [convert()].
#' @export
conversion_options <- function(
  heading_style = NULL,
  list_indent_type = NULL,
  list_indent_width = NULL,
  bullets = NULL,
  strong_em_symbol = NULL,
  escape_asterisks = NULL,
  escape_underscores = NULL,
  escape_misc = NULL,
  escape_ascii = NULL,
  code_language = NULL,
  autolinks = NULL,
  default_title = NULL,
  br_in_tables = NULL,
  compact_tables = NULL,
  highlight_style = NULL,
  extract_metadata = NULL,
  whitespace_mode = NULL,
  strip_newlines = NULL,
  wrap = NULL,
  wrap_width = NULL,
  convert_as_inline = NULL,
  sub_symbol = NULL,
  sup_symbol = NULL,
  newline_style = NULL,
  code_block_style = NULL,
  keep_inline_images_in = NULL,
  preprocessing = NULL,
  encoding = NULL,
  debug = NULL,
  strip_tags = NULL,
  preserve_tags = NULL,
  skip_images = NULL,
  url_escape_style = NULL,
  link_style = NULL,
  output_format = NULL,
  include_document_structure = NULL,
  extract_images = NULL,
  max_image_size = NULL,
  capture_svg = NULL,
  infer_dimensions = NULL,
  max_depth = NULL,
  exclude_selectors = NULL,
  tier_strategy = NULL,
  visitor = NULL
) {
  opts <- list()
  if (!is.null(heading_style)) opts$heading_style <- heading_style
  if (!is.null(list_indent_type)) opts$list_indent_type <- list_indent_type
  if (!is.null(list_indent_width)) opts$list_indent_width <- as.integer(list_indent_width)
  if (!is.null(bullets)) opts$bullets <- bullets
  if (!is.null(strong_em_symbol)) opts$strong_em_symbol <- strong_em_symbol
  if (!is.null(escape_asterisks)) opts$escape_asterisks <- escape_asterisks
  if (!is.null(escape_underscores)) opts$escape_underscores <- escape_underscores
  if (!is.null(escape_misc)) opts$escape_misc <- escape_misc
  if (!is.null(escape_ascii)) opts$escape_ascii <- escape_ascii
  if (!is.null(code_language)) opts$code_language <- code_language
  if (!is.null(autolinks)) opts$autolinks <- autolinks
  if (!is.null(default_title)) opts$default_title <- default_title
  if (!is.null(br_in_tables)) opts$br_in_tables <- br_in_tables
  if (!is.null(compact_tables)) opts$compact_tables <- compact_tables
  if (!is.null(highlight_style)) opts$highlight_style <- highlight_style
  if (!is.null(extract_metadata)) opts$extract_metadata <- extract_metadata
  if (!is.null(whitespace_mode)) opts$whitespace_mode <- whitespace_mode
  if (!is.null(strip_newlines)) opts$strip_newlines <- strip_newlines
  if (!is.null(wrap)) opts$wrap <- wrap
  if (!is.null(wrap_width)) opts$wrap_width <- as.integer(wrap_width)
  if (!is.null(convert_as_inline)) opts$convert_as_inline <- convert_as_inline
  if (!is.null(sub_symbol)) opts$sub_symbol <- sub_symbol
  if (!is.null(sup_symbol)) opts$sup_symbol <- sup_symbol
  if (!is.null(newline_style)) opts$newline_style <- newline_style
  if (!is.null(code_block_style)) opts$code_block_style <- code_block_style
  if (!is.null(keep_inline_images_in)) opts$keep_inline_images_in <- keep_inline_images_in
  if (!is.null(preprocessing)) opts$preprocessing <- preprocessing
  if (!is.null(encoding)) opts$encoding <- encoding
  if (!is.null(debug)) opts$debug <- debug
  if (!is.null(strip_tags)) opts$strip_tags <- strip_tags
  if (!is.null(preserve_tags)) opts$preserve_tags <- preserve_tags
  if (!is.null(skip_images)) opts$skip_images <- skip_images
  if (!is.null(url_escape_style)) opts$url_escape_style <- url_escape_style
  if (!is.null(link_style)) opts$link_style <- link_style
  if (!is.null(output_format)) opts$output_format <- output_format
  if (!is.null(include_document_structure)) opts$include_document_structure <- include_document_structure
  if (!is.null(extract_images)) opts$extract_images <- extract_images
  if (!is.null(max_image_size)) opts$max_image_size <- as.integer(max_image_size)
  if (!is.null(capture_svg)) opts$capture_svg <- capture_svg
  if (!is.null(infer_dimensions)) opts$infer_dimensions <- infer_dimensions
  if (!is.null(max_depth)) opts$max_depth <- as.integer(max_depth)
  if (!is.null(exclude_selectors)) opts$exclude_selectors <- exclude_selectors
  if (!is.null(tier_strategy)) opts$tier_strategy <- tier_strategy
  if (!is.null(visitor)) opts$visitor <- visitor
  opts
}
